### PR TITLE
chore(core): users_entity subtable attributes are moved to metadata 

### PR DESCRIPTION
**Pull request by jdalsem**
- [x] add check for duplicate usernames
- [x] add check for emptying required/important metadata

### CODE REVIEW

**Comment by hypeJunction** [comment]
Let's discuss this before merging. I like that it's all red, but I don't think username and password hash belongs in metadata. First of all, metadata storage allows arrays, so we need to ensure these values are always scalar. Perhaps private settings table is a better place for these values, but I am not sure.

**Comment by jdalsem** [comment]
the red is caused by the fact this PR is the last subtable related PR and also removes the attribute related tests and functions.

Username is indeed a point of discussion. We could make a magic setter/getter to handle this maybe. Don't think private settings is better perse..

Username is also required to be unique so we need something special for that field anyways

**Comment by mrclay** [comment]
Both these are going to bite us hard when we move almost all critical content to MD:

- [ ] #4268 Updating metadata is not thread-safe
- [ ] #10636 User delete causes owned metadata deletes

**Comment by hypeJunction** [comment]
Yet again, we should just require `ElggEntity::save()` to persist metadata. Solves a bunch of issues, will be a pain to upgrade though.

**Comment by mrclay** [comment]
> users_entity table and its data

Its data will be put in danger by this PR. We've been able to ignore a serious data integrity issue because relatively little depended on it, and this is a huge step in changing that.

File: engine/classes/Elgg/Database/EntityTable.php [hunk 1]
```diff
@@ -730,46 +729,6 @@ public function getEntities(array $options = []) {
 	 * @return array
 	 */
 	protected function autoJoinTables(array $options) {
```
Comment (mrclay) [inline engine/classes/Elgg/Database/EntityTable.php]: We just can't give this function up ;)

Comment (jdalsem) [reply engine/classes/Elgg/Database/EntityTable.php]: yep

**Comment by jdalsem** [comment]
user

> Its data will be put in danger by this PR. We've been able to ignore a serious data integrity issue because relatively little depended on it, and this is a huge step in changing that.

it is not that big of a deal i think... user metadata (like profile data) is changing much more often than attributes (like title/name/username) and in my whole Elgg live i have never ran into this theoretical issue. I do not think we are making the issue much more a problem with this PR. If the data integrity is in danger, i think the importance of regular metadata is the same as attributes.

Lets just put the issue on the roadmap for 3.0 and do some fixing. I think there are already some quick wins to do on metadata so it does updates more often than delete/insert (which is a big cause of the thread safety).

**Comment by hypeJunction** [comment]
> thank you for reporting all kinds of related ideas or issues

These are not related ideas. We are trying to identify potential issues with migrating critical user attributes to metadata. Objects, groups and sites attributes are not controversial, but user attributes are. 

Remember that users are registered without a logged in session. Metadata is subject to capability checks, so right there you have a potential for data loss, incomplete user entities etc.

We don't have basic sanity checks for attributes, so you may end up writing two usernames to a user, what happens then?

**Comment by jdalsem** [comment]
> We don't have basic sanity checks for attributes, so you may end up writing two usernames to a user, what happens then?

a check for that is underway

**Comment by hypeJunction** [comment]
`php
// We need a define to ensure that certain metadata is always scalar
$user->username = ['username1', 'username2'];

// We need to make sure required metadata is not overwritten
$user->username = '';
$user->password_hash = '';
$user->email = '';

// We need to make sure changes to attribute-metadata is validated
$user->email = 'invalid.@email';

// We need to make sure enumerated metadata properties are validated
$user->admin = 'whatever';
```

* What happens to all the code that depends on `original_attributes` in `update` events?

**Comment by jdalsem** [comment]
> We need a define to ensure that certain metadata is always scalar
> $user->username = ['username1', 'username2'];

agreed

> // We need to make sure required metadata is not overwritten
> $user->username = '';
> $user->password_hash = '';
> $user->email = '';

Agreed, is already done for password_hash


> // We need to make sure changes to attribute-metadata is validated
> $user->email = 'invalid.@email';

Disagree as part of this PR as record validation is also not done previously for email

> // We need to make sure enumerated metadata properties are validated
> $user->admin = 'whatever';

Disagree, don't think that this is that important.

**Comment by hypeJunction** [comment]
I am working on a registration workflow for a project now, and what would really be helpful if we could do a dry run with user attributes without actually persisting changes. This would allow to validate all props without having to wait for something to throw.

File: actions/admin/user/ban.php [hunk 2]
```diff
@@ -2,8 +2,7 @@
 /**
  * Bans a user.
  *
- * User entities are banned by setting the 'banned' column
- * to 'yes' in the users_entity table.
+ * User entities are banned by setting the 'banned' metadata
```
Comment (jeabakker) [inline actions/admin/user/ban.php]: to 'yes'

File: engine/classes/Elgg/Database/UsersTable.php [hunk 3]
```diff
@@ -549,34 +360,9 @@ public function setLastAction(ElggUser $user) {
...
-		";
-
-		$params = [
-			':last_action' => $time,
-			':guid' => (int) $user->guid,
-		];
-
 		// these writes actually work, we just type hint read-only.
 		$user->prev_last_action = $user->last_action;
 		$user->last_action = $time;
```
Comment (jeabakker) [inline engine/classes/Elgg/Database/UsersTable.php]: since this is an Entity attribute this still needs to be saved to the database

File: engine/lib/admin.php [hunk 4]
```diff
@@ -29,25 +29,10 @@
...
-		$options['wheres'][] = "u.admin = 'yes'";
-	} else {
-		$options['wheres'][] = "u.admin = 'yes'";
-	}
+	$options['type'] = 'user';
+	$options['metadata_name_value_pairs'] = elgg_extract('metadata_name_value_pairs', $options, []);
+	
+	$options['metadata_name_value_pairs'][] = ['admin' => 'yes'];
 
 	return elgg_get_entities($options);
```
Comment (jeabakker) [inline engine/lib/admin.php]: `elgg_get_entities_from_metadata`

File: engine/lib/deprecated-3.0.php [hunk 5]
```diff
@@ -185,3 +185,84 @@ function elgg_get_entities_from_attributes(array $options = []) {
...
+ *
+ * @param int    $user_guid The user guid
+ * @param string $reason    A reason
+ *
+ * @return bool
+ *
+ * @deprecated Use \ElggUser->ban()
+ */
+function ban_user($user_guid, $reason = "") {
+	elgg_deprecated_notice(__FUNCTION__ . ' is deprecated. Use \ElggUser::ban() a.', '3.0');
```
Comment (jeabakker) [inline engine/lib/deprecated-3.0.php]: why at the end 'a.'

File: engine/lib/deprecated-3.0.php [hunk 6]
```diff
@@ -185,3 +185,84 @@ function elgg_get_entities_from_attributes(array $options = []) {
...
+ * Unban a user.
+ *
+ * @param int $user_guid Unban a user.
+ *
+ * @return bool
+ *
+ * @deprecated Use \ElggUser->unban()
+ */
+function unban_user($user_guid) {
+	elgg_deprecated_notice(__FUNCTION__ . ' is deprecated. Use \ElggUser::unban() a.', '3.0');
```
Comment (jeabakker) [inline engine/lib/deprecated-3.0.php]: why at the end 'a.'

File: engine/lib/deprecated-3.0.php [hunk 7]
```diff
@@ -185,3 +185,84 @@ function elgg_get_entities_from_attributes(array $options = []) {
...
+ * Makes user $guid an admin.
+ *
+ * @param int $user_guid User guid
+ *
+ * @return bool
+ *
+ * @deprecated Use \ElggUser->makeAdmin()
+ */
+function make_user_admin($user_guid) {
+	elgg_deprecated_notice(__FUNCTION__ . ' is deprecated. Use \ElggUser::makeAdmin() a.', '3.0');
```
Comment (jeabakker) [inline engine/lib/deprecated-3.0.php]: why at the end 'a.'

File: engine/lib/deprecated-3.0.php [hunk 8]
```diff
@@ -185,3 +185,84 @@ function elgg_get_entities_from_attributes(array $options = []) {
...
+ * Removes user $guid's admin flag.
+ *
+ * @param int $user_guid User GUID
+ *
+ * @return bool
+ *
+ * @deprecated Use \ElggUser->removeAdmin()
+ */
+function remove_user_admin($user_guid) {
+	elgg_deprecated_notice(__FUNCTION__ . ' is deprecated. Use \ElggUser::removeAdmin() a.', '3.0');
```
Comment (jeabakker) [inline engine/lib/deprecated-3.0.php]: why at the end 'a.'

File: engine/tests/classes/Elgg/Mocks/Database/EntityTable.php [hunk 9]
```diff
@@ -80,62 +80,53 @@ public function setup($guid, $type, $subtype, array $attributes = []) {
 			$subtype = $this->subtype_table->getSubtype($subtype_id);
 		}
 
+		// don't let these to be overriden by attributes
 		$attributes['guid'] = $guid;
-		$attributes['type'] = $type;
-		$attributes['subtype'] = $subtype_id;
+ 		$attributes['type'] = $type;
+ 		$attributes['subtype'] = $subtype_id;
```
Comment (jeabakker) [inline engine/tests/classes/Elgg/Mocks/Database/EntityTable.php]: ?? added space?

File: engine/tests/classes/Elgg/Mocks/Database/MetadataTable.php [hunk 10]
```diff
@@ -61,7 +62,7 @@ public function create($entity_guid, $name, $value, $value_type = '', $owner_gui
 		];
 
 		$this->rows[$id] = $row;
-
+
```
Comment (jeabakker) [inline engine/tests/classes/Elgg/Mocks/Database/MetadataTable.php]: no changes in this file, try to exclude it from the commit

File: engine/tests/phpunit/unit/Elgg/EmailUnitTest.php [hunk 11]
```diff
@@ -18,16 +18,16 @@ public function down() {
...
 	public function testFactoryFromElggUser() {
 
-		$from = $this->createUser([
+		$from = $this->createUser([], [
 			'email' => 'from@elgg.org',
 			'name' => 'From',
 		]);
-		
-		$to = $this->createUser([
+
```
Comment (jeabakker) [inline engine/tests/phpunit/unit/Elgg/EmailUnitTest.php]: lots of additional tabs

File: mod/garbagecollector/start.php [hunk 12]
```diff
@@ -100,9 +100,7 @@ function garbagecollector_optimize_table($table) {
 function garbagecollector_entities() {
 	$dbprefix = elgg_get_config('dbprefix');
 
-	$tables = [
-		'user' => 'users_entity',
-	];
+	$tables = [];
```
Comment (jeabakker) [inline mod/garbagecollector/start.php]: is the foreach below this still functional/needed

File: mod/notifications/views/default/notifications/subscriptions/users.php [hunk 13]
```diff
@@ -12,18 +12,23 @@
 
 // Returns a list of all friends, as well as anyone else who the user is subscribed to
 $dbprefix = elgg_get_config('dbprefix');
-$options = [
+
+echo elgg_list_entities([
```
Comment (jeabakker) [inline mod/notifications/views/default/notifications/subscriptions/users.php]: `elgg_list_entities_from_metadata`

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/Database/UsersTable.php (hunk 3 above)
Comment (jdalsem) [reply engine/classes/Elgg/Database/UsersTable.php]: done

**Review by jeabakker — APPROVED** [review]

File: engine/classes/ElggUser.php [hunk 14]
```diff
@@ -47,36 +45,31 @@ public function getLanguage($fallback = null) {
...
+				}
 				break;
-
-			default:
-				parent::__set($name, $value);
+			case 'username':
+				if (!validate_username($value)) {
+					throw new \InvalidParameterException("Username is not a valid username");
+				}
+				$existing_user = get_user_by_username($value);
```
Comment (hypeJunction) [inline engine/classes/ElggUser.php]: Shouldn't this check disabled entities as well?

File: engine/classes/ElggUser.php [hunk 15]
```diff
@@ -99,8 +92,23 @@ public function getURL() {
...
+		}
+		
+		if (!elgg_trigger_event('ban', 'user', $this)) {
+			return false;
+		}
+
+		$this->ban_reason = $reason;
+		$this->banned = 'yes';
+				
+		_elgg_invalidate_cache_for_entity($this->guid);
```
Comment (hypeJunction) [inline engine/classes/ElggUser.php]: The probably is not necessary any more

File: engine/classes/ElggUser.php [hunk 16]
```diff
@@ -127,12 +150,11 @@ public function isBanned() {
 	 * @return bool
 	 */
 	public function isAdmin() {
-
-		// for backward compatibility we need to pull this directly
-		// from the attributes instead of using the magic methods.
-		// this can be removed in 1.9
-		// return $this->admin == 'yes';
-		return $this->attributes['admin'] == 'yes';
+		$ia = elgg_set_ignore_access(true);
```
Comment (hypeJunction) [inline engine/classes/ElggUser.php]: Why is this needed?

File: engine/classes/ElggUser.php [hunk 17]
```diff
@@ -146,14 +168,15 @@ public function makeAdmin() {
...
-		if ($this->guid && !make_user_admin($this->guid)) {
+		if (!elgg_trigger_event('make_admin', 'user', $this)) {
 			return false;
 		}
 
-		// need to manually set attributes since they've already been loaded.
-		$this->attributes['admin'] = 'yes';
+		$this->admin = 'yes';
 
+		_elgg_invalidate_cache_for_entity($this->guid);
```
Comment (hypeJunction) [inline engine/classes/ElggUser.php]: I don't think this is necessary any more

File: engine/lib/deprecated-3.0.php [hunk 18]
```diff
@@ -185,3 +185,84 @@ function elgg_get_entities_from_attributes(array $options = []) {
...
+ * Ban a user
+ *
+ * @param int    $user_guid The user guid
+ * @param string $reason    A reason
+ *
+ * @return bool
+ *
+ * @deprecated Use \ElggUser->ban()
+ */
+function ban_user($user_guid, $reason = "") {
```
Comment (hypeJunction) [inline engine/lib/deprecated-3.0.php]: Just get rid of these

File: engine/lib/sessions.php [hunk 19]
```diff
@@ -72,26 +72,12 @@ function elgg_is_admin_user($user_guid) {
 	
 	$user_guid = (int) $user_guid;
 
-	$current_user = elgg_get_logged_in_user_entity();
```
Comment (hypeJunction) [inline engine/lib/sessions.php]: Let's just get rid of this function. Useless API bloat

File: engine/lib/deprecated-3.0.php (hunk 18 above)
Comment (jdalsem) [reply engine/lib/deprecated-3.0.php]: see my new PR

File: engine/classes/ElggUser.php (hunk 14 above)
Comment (jdalsem) [reply engine/classes/ElggUser.php]: good point... need to fix that

Comment (hypeJunction) [reply engine/classes/ElggUser.php]: Let's add public API to check username availability. I keep reinventing the wheel every time
